src/Compiler/Checking/TailCallChecks.fs

```diff
@@ -756,7 +756,25 @@ let CheckModuleBinding cenv (isRec: bool) (TBind _ as bind) =
                 | Expr.Lambda(bodyExpr = bodyExpr) -> checkTailCall insideSubBindingOrTry bodyExpr
                 | Expr.DebugPoint(_debugPointAtLeafExpr, expr) -> checkTailCall insideSubBindingOrTry expr
                 | Expr.Let(binding = binding; bodyExpr = bodyExpr) ->
-                    checkTailCall true binding.Expr
+                    let isAsyncContinuation =
+                        match bodyExpr with
+                        | Expr.App(funcExpr = Expr.Val(valRef = valRef)) ->
```

Comment:
Are we 100% sure resulting code will be tail recursive?
```fsharp
namespace N
module M =
    [<TailCall>]
    let rec f (g: bool Async) =
        async {
            match! g with
            | false -> ()
            | true -> return! f g
        }

    [<EntryPoint>]
    let main _ =
        f ( async { return true } ) |> Async.RunSynchronously
        0
```

Will result in chained calls of various generated closure classes:
```csharp
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Microsoft.FSharp.Control;
using Microsoft.FSharp.Core;

namespace N;

[CompilationMapping(SourceConstructFlags.Module)]
public static class M
{
    [Serializable]
    internal sealed class f@6-1 : FSharpFunc<bool, FSharpAsync<Unit>>
    {
        public FSharpAsync<bool> g;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6-1(FSharpAsync<bool> g)
        {
            this.g = g;
        }

        public override FSharpAsync<Unit> Invoke(bool _arg1)
        {
            if (!_arg1)
            {
                return ExtraTopLevelOperators.DefaultAsyncBuilder.Zero();
            }
            return f(g);
        }
    }

    [Serializable]
    internal sealed class f@6-2 : FSharpFunc<AsyncActivation<Unit>, AsyncReturn>
    {
        public FSharpAsync<bool> g;

        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        [CompilerGenerated]
        [DebuggerNonUserCode]
        public FSharpFunc<bool, FSharpAsync<Unit>> binder;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6-2(FSharpAsync<bool> g, FSharpFunc<bool, FSharpAsync<Unit>> binder)
        {
            this.g = g;
            this.binder = binder;
        }

        public override AsyncReturn Invoke(AsyncActivation<Unit> ctxt)
        {
            return AsyncPrimitives.Bind(ctxt, g, binder);
        }
    }

    [Serializable]
    internal sealed class f@6 : FSharpFunc<Unit, FSharpAsync<Unit>>
    {
        public FSharpAsync<bool> g;

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal f@6(FSharpAsync<bool> g)
        {
            this.g = g;
        }

        public override FSharpAsync<Unit> Invoke(Unit unitVar)
        {
            FSharpFunc<bool, FSharpAsync<Unit>> binder = new f@6-1(g);
            return AsyncPrimitives.MakeAsync(new f@6-2(g, binder));
        }
    }

    [Serializable]
    internal sealed class main@13-1 : FSharpFunc<AsyncActivation<bool>, AsyncReturn>
    {
        internal static readonly main@13-1 @_instance = new main@13-1();

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal main@13-1()
        {
        }

        public override AsyncReturn Invoke(AsyncActivation<bool> ctxt)
        {
            return AsyncActivation<bool>.Success(ctxt, result: true);
        }
    }

    [Serializable]
    internal sealed class main@13 : FSharpFunc<Unit, FSharpAsync<bool>>
    {
        internal static readonly main@13 @_instance = new main@13();

        [CompilerGenerated]
        [DebuggerNonUserCode]
        internal main@13()
        {
        }

        public override FSharpAsync<bool> Invoke(Unit unitVar)
        {
            return AsyncPrimitives.MakeAsync(main@13-1.@_instance);
        }
    }

    [TailCall]
    public static FSharpAsync<Unit> f(FSharpAsync<bool> g)
    {
        return ExtraTopLevelOperators.DefaultAsyncBuilder.Delay(new f@6(g));
    }

    [EntryPoint]
    public static int main(string[] _arg1)
    {
        FSharpAsync<Unit> computation = f(ExtraTopLevelOperators.DefaultAsyncBuilder.Delay(main@13.@_instance));
        FSharpAsync.RunSynchronously(computation, null, null);
        return 0;
    }
}
```

Reply:
I suppose it is since we call f -> f@6 -> f@6-2 -> 6@6-1 -> f or something.

Reply:
Yes, from what I see, it's tail recursive.